keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,276 @@
+  Shared quota cannot be consumed by requests queued at lower priorities
+  than the bucket's.
+  - *Weight*: only works for shared quota. The higher weight a bucket has,
+  the more chance the requests hitting the buckets can be executed from the
+  queue.
+  - *Priority*: a fixed/preset set of priorities ranged 1...N. The lower
+  number means higher priority. Each priority has one or more buckets providing
+  quotas and buckets from higher priority can "borrow" shared quotas from
+  the lowers.
+- *Extra Shared Quota / ESQ*: global shared quota provided by a logical
+bucket at a virtual "lowest" priority. Requests doesn't match any priority
+will be queued in this logical "lowest" priority until ESQ distributes quota
+to it. ESQ can also be consumed by requests from all priorities. Generally,
+ESQ is a default common concurrency in the system and it's supposed to be
+auto-tunable by a moving average observed by the system. Note that requests
+queued in this extra bandwidth are popped strictly in a FIFO order.
+
+##### Bandwidth Goal
+
+We can define __Shared Bandwidth__ as a goal for priority band __p__, and
+we have N fixed priority:
+
+- *Defined Shared Bandwidth(p) / DSB(p)*: sum of shared quotas provided by
+buckets at Priority p.
+- *System Total Shared Bandwidth / STSB*: sum\[ DSB(i) over 1..N \] + ESQ
+- *Maximum Usable Shared Bandwidth(p) / MUSB*:
+  - if p == 1: STSB
+  - else: STSB - sum\[ STSB - sum\[MUSB(i) over 1...p-1 \]
+
+Generally, the system's overall shared bandwidth consists of a sum of shared
+quotas defined by all buckets and an auto-tunable extra shared bandwidth. In
+this so-called shared bandwidth, a request w/ higher priority can always be
+executed prior to the requests w/ lower priority. Which is, at worst case,
+when requests from top priority completely occupies the shared bandwidth,
+then requests from lower bands can only be executed under its reserved quota.
+And those "wild" request doesn't match any buckets will be rejected.
+
+##### Overview
+
+Within each priority band, we have a set of buckets matching the priority and
+a WRR request queue. Request that missed reserved quota will be queued into
+the WRR queue w/ the weight of its bucket. Note that reserving quota of
+buckets at priority N can also re-claim requests queued at N.
+
+All requests will be pre-processed before it's actually executed by 3 stages:
+
+0. We find a matched bucket by applying BucketBinding API.
+1. *Reserving*: We check if the bucket's reserved quota. If it's sufficient,
+the request will consume one reserved quota from the bucket and be executed
+immediately.
+2. *Prioritizing*: We classify and enqueue the request into a priority band.
+Requests at higher priority will __always__ be executed prior to requests at lower.
+3. *Distributing*: We poll requests from each priority band's queue by
+[Weighted Round Robin](https://en.wikipedia.org/wiki/Weighted_round_robin)
```

Comment:
One of the properties we want is fairness. Mike and I seem to have both (independently?) thought of multiple queues & shuffle sharding. Is there any fairness mechanism in this design?

Reply:
yes, in a priority band, we have a set of queues and these queues' identifier is the name of the bucket referenced by the matched binding rule. so when the requests arrives at the apiserver, they will be sharded first by priority, then by bucket name into plain FIFO queues. the drainer will poll requests out of the queues by applying WRR algorithm. simply put, the fairness are guaranteed in each priority band.

Reply:
> yes, in a priority band, we have a set of queues and these queues' identifier is the name of the bucket referenced by the matched binding rule. so when the requests arrives at the apiserver, they will be sharded first by priority, then by bucket name into plain FIFO queues. the drainer will poll requests out of the queues by applying WRR algorithm. simply put, the fairness are guaranteed in each priority band.

I think the question is whether a high priority writer can starve the lower priority.  Based on my read, it could starve the shared, but it would not starve the reserved.

Reply:
> Based on my read, it could starve the shared, but it would not starve the reserved.

exactly